packages/amplify-provider-awscloudformation/src/iterative-deployment/stack-progress-printer.ts

```diff
@@ -1,23 +1,123 @@
-import { StackEvent, StackEvents } from 'aws-sdk/clients/cloudformation';
-import { IStackProgressPrinter } from './stack-event-monitor';
-import columnify from 'columnify';
+import { MultiProgressBar } from 'amplify-prompts';
+import type { StackEvent, StackEvents } from 'aws-sdk/clients/cloudformation';
 import chalk from 'chalk';
-import ora, { Ora } from 'ora';
-
-const CFN_SUCCESS_STATUS = ['UPDATE_COMPLETE', 'CREATE_COMPLETE', 'DELETE_COMPLETE', 'DELETE_SKIPPED'];
+import columnify from 'columnify';
+import {
+  CFN_SUCCESS_STATUS, CNF_ERROR_STATUS, createItemFormatter, createProgressBarFormatter, EventMap,
+} from '../utils/progress-bar-helpers';
+import { IStackProgressPrinter } from './stack-event-monitor';
 
-const CNF_ERROR_STATUS = ['CREATE_FAILED', 'DELETE_FAILED', 'UPDATE_FAILED'];
+/**
+ * Iterative deployment stack printer.
+ */
 export class StackProgressPrinter implements IStackProgressPrinter {
   private events: StackEvents = [];
-  private isRunning: boolean = true;
-  private spinner: Ora = ora('Deploying');
-  addActivity = (event: StackEvent) => {
+  private progressBars: MultiProgressBar;
+  private eventMap: EventMap;
+  private categoriesPrinted: string[] = [];
+
+  constructor(eventMap: EventMap) {
+    // Initialize the progress bar and event map.
+    this.progressBars = new MultiProgressBar({
+      progressBarFormatter: createProgressBarFormatter,
+      itemFormatter: createItemFormatter,
+      loneWolf: false,
```

Comment:
what's lone wolf?

Reply:
https://github.com/aws-amplify/amplify-cli/blob/dev/packages/amplify-prompts/src/progressbars/progressbar.ts#L10

Reply:
That's surprising.
Why do we pass that flag to multi progress bar then? Isn't multi bar an aggregate of non-lone-wolfs by definition?
Why bars can't detect this themselves through child-parent relationship? (assuming that they form a tree).

Reply:
It's due to a break down in abstractions. When a progressbar is part of a multibar, the multibar handles the rendering. Added a backlog item to refactor this into a cleaner abstraction.